### PR TITLE
DMP-4198: Populate data_anonymisation table

### PR DESCRIPTION
**Pull request by Ben-Edwards-cgi**
### Change description ###
As part of the v72 data model, a data_anonymisation table has been introduced. This is used to track the details for when data is anonymised as part of case expiry or manual obfuscation. 

When a case is expired, the 'requested by' and 'approved by' values are set to the user account run by the case expiry job. The dates should be set to the current date/time. This will cover all events and transcription comments that are anonymised as part of case expiry. 

When an event is manually anonymised, the anonymisation is recorded as a manual request with the requested by and approved by fields both set to the user that performed the anonymisation. This will change to different users if DMP-4116 gets played. 
h4. Acceptance criteria

*AC1: Case expiry job runs*

GIVEN

an case has exceeded it's retention date

AND 

it is associated with events and transcription requests (with comments)

WHEN

the case expiry job runs and the data is anonymised

THEN

record the event and transcription comment in the data_anonymisation table

*AC2: Manual event text obfuscation*

GIVEN

an event is manually obfuscated

WHEN

the request is received

THEN

record the event detail in the data_anonymisation table

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[X] No
```

### CODE REVIEW

File: src/test/java/uk/gov/hmcts/darts/common/service/impl/DataAnonymisationServiceImplTest.java [hunk 1]
```diff
@@ -95,27 +105,39 @@ void eventEntityAnonymiseNotUpdatedAsNotAllCasesExpiredAndOnlyAnonymiseIfAllCase
...
         assertThat(eventEntity.isDataAnonymised()).isFalse();
         verify(eventService).allAssociatedCasesAnonymised(eventEntity);
         verify(eventService, never()).saveEvent(eventEntity);
     }
 
-    @Test
-    void positiveEventEntityAnonymiseUpdatedAsAllCasesExpiredAndOnlyAnonymiseIfAllCasesExpiredIsTrue() {
+    @ParameterizedTest(name = "Event should be anonymised if all assocaited cases are anonymised. (isManuallyRequested = {0})")
+    @ValueSource(booleans = {true, false})
+    void eventEntityAnonymiseUpdatedAsAllCasesExpiredAndOnlyAnonymiseIfAllCasesExpiredIsTrue(boolean isManuallyRequested) {
```
Comment (davet1985) [inline src/test/java/uk/gov/hmcts/darts/common/service/impl/DataAnonymisationServiceImplTest.java:117]: Can you amend the name to use the new convention please

File: src/test/java/uk/gov/hmcts/darts/common/service/impl/DataAnonymisationServiceImplTest.java [hunk 2]
```diff
@@ -186,23 +208,29 @@ void positiveAnonymiseProsecutorEntity() {
         assertLastModifiedByAndAt(prosecutorEntity, userAccount);
     }
 
-    @Test
-    void positiveAnonymiseTranscriptionCommentEntity() {
+    @ParameterizedTest(name = "Anonymise TranscriptionCommentEntity with isManuallyRequested = {0}")
+    @ValueSource(booleans = {true, false})
+    void anonymiseTranscriptionCommentEntity(boolean isManuallyRequested) {
```
Comment (davet1985) [inline src/test/java/uk/gov/hmcts/darts/common/service/impl/DataAnonymisationServiceImplTest.java:213]: Can you amend the name to use the new convention please

File: src/test/java/uk/gov/hmcts/darts/common/service/impl/DataAnonymisationServiceImplTest.java [hunk 3]
```diff
@@ -186,23 +208,29 @@ void positiveAnonymiseProsecutorEntity() {
...
         assertLastModifiedByAndAt(transcriptionCommentEntity, userAccount);
+        verify(dataAnonymisationService).registerDataAnonymisation(userAccount, transcriptionCommentEntity, isManuallyRequested);
     }
 
 
-    @Test
-    void assertPositiveAnonymiseCourtCaseEntity() {
+    @ParameterizedTest(name = "Anonymise CourtCase with isManuallyRequested = {0}")
+    @ValueSource(booleans = {true, false})
+    void assertPositiveAnonymiseCourtCaseEntity(boolean isManuallyRequested) {
```
Comment (davet1985) [inline src/test/java/uk/gov/hmcts/darts/common/service/impl/DataAnonymisationServiceImplTest.java:233]: Can you amend the name to use the new convention please

File: src/test/java/uk/gov/hmcts/darts/common/service/impl/DataAnonymisationServiceImplTest.java [hunk 4]
```diff
@@ -225,44 +253,51 @@ void assertPositiveAnonymiseCourtCaseEntity() {
...
+        verify(eventService).allAssociatedCasesAnonymised(eventEntity1);
+        verify(eventService).allAssociatedCasesAnonymised(eventEntity2);
     }
 
 
-    @Test
-    void positiveAnonymiseTranscriptionEntity() {
+    @ParameterizedTest(name = "Anonymise Transcription with isManuallyRequested = {0}")
+    @ValueSource(booleans = {true, false})
+    void anonymiseTranscriptionEntity(boolean isManuallyRequested) {
```
Comment (davet1985) [inline src/test/java/uk/gov/hmcts/darts/common/service/impl/DataAnonymisationServiceImplTest.java:300]: Can you amend the name to use the new convention please

File: src/test/java/uk/gov/hmcts/darts/common/service/impl/DataAnonymisationServiceImplTest.java [hunk 5]
```diff
@@ -273,22 +308,23 @@ void positiveAnonymiseTranscriptionEntity() {
...
-        verify(dataAnonymisationService, times(1)).anonymiseTranscriptionWorkflowEntity(transcriptionWorkflowEntity2);
+        verify(dataAnonymisationService).anonymiseTranscriptionWorkflowEntity(transcriptionWorkflowEntity1);
+        verify(dataAnonymisationService).anonymiseTranscriptionWorkflowEntity(transcriptionWorkflowEntity2);
     }
 
-    @Test
-    void positiveAnonymiseHearingEntity() {
+    @ParameterizedTest(name = "Anonymise hearing with isManuallyRequested = {0}")
+    @ValueSource(booleans = {true, false})
+    void anonymiseHearingEntity(boolean isManuallyRequested) {
```
Comment (davet1985) [inline src/test/java/uk/gov/hmcts/darts/common/service/impl/DataAnonymisationServiceImplTest.java:327]: Can you amend the name to use the new convention please

File: src/test/java/uk/gov/hmcts/darts/common/service/impl/DataAnonymisationServiceImplTest.java [hunk 6]
```diff
@@ -419,32 +455,33 @@ void positiveTidyUpTransformedMediaEntities() {
...
-        verify(dataAnonymisationService, times(1))
+        verify(dataAnonymisationService)
             .deleteTransformedMediaEntity(transformedMediaEntity5);
     }
 
-    @Test
-    void positiveAnonymiseEventByIds() {
+    @ParameterizedTest(name = "Anonymise event by ids with isManuallyRequested = {0}")
+    @ValueSource(booleans = {true, false})
+    void anonymiseEventByIds(boolean isManuallyRequested) {
```
Comment (davet1985) [inline src/test/java/uk/gov/hmcts/darts/common/service/impl/DataAnonymisationServiceImplTest.java:484]: Can you amend the name to use the new convention please

File: src/test/java/uk/gov/hmcts/darts/common/service/impl/DataAnonymisationServiceImplTest.java [hunk 7]
```diff
@@ -455,21 +492,70 @@ void positiveAnonymiseEventByIds() {
...
+        verify(eventService).saveEvent(event3);
         verifyNoMoreInteractions(eventService);
+        verify(dataAnonymisationService).registerDataAnonymisation(userAccount, event1, isManuallyRequested);
+        verify(dataAnonymisationService).registerDataAnonymisation(userAccount, event2, isManuallyRequested);
+        verify(dataAnonymisationService).registerDataAnonymisation(userAccount, event3, isManuallyRequested);
+    }
+
+    @ParameterizedTest(name = "Register data anonymisation event with isManuallyRequested = {0}")
+    @ValueSource(booleans = {true, false})
+    void registerDataAnonymisationEvent(boolean isManuallyRequested) {
```
Comment (davet1985) [inline src/test/java/uk/gov/hmcts/darts/common/service/impl/DataAnonymisationServiceImplTest.java:520]: Can you amend the name to use the new convention please

File: src/test/java/uk/gov/hmcts/darts/common/service/impl/DataAnonymisationServiceImplTest.java [hunk 8]
```diff
@@ -455,21 +492,70 @@ void positiveAnonymiseEventByIds() {
...
+        assertThat(dataAnonymisationEntity.getRequestedBy()).isEqualTo(userAccount);
+        assertThat(dataAnonymisationEntity.getRequestedTs()).isEqualTo(currentTime);
+        assertThat(dataAnonymisationEntity.getApprovedBy()).isEqualTo(userAccount);
+        assertThat(dataAnonymisationEntity.getApprovedTs()).isEqualTo(currentTime);
+        verify(currentTimeHelper, times(1)).currentOffsetDateTime();
+    }
+
+    @ParameterizedTest(name = "Register data anonymisation transcription comment with isManuallyRequested = {0}")
+    @ValueSource(booleans = {true, false})
+    void registerDataAnonymisationTranscriptionComment(boolean isManuallyRequested) {
```
Comment (davet1985) [inline src/test/java/uk/gov/hmcts/darts/common/service/impl/DataAnonymisationServiceImplTest.java:542]: Can you amend the name to use the new convention please

**Review by davet1985 — CHANGES_REQUESTED** [review]
Just test names to update I think 😉

File: src/test/java/uk/gov/hmcts/darts/common/service/impl/DataAnonymisationServiceImplTest.java (hunk 1 above)
Comment (Ben-Edwards-cgi) [reply src/test/java/uk/gov/hmcts/darts/common/service/impl/DataAnonymisationServiceImplTest.java]: Updated

File: src/test/java/uk/gov/hmcts/darts/common/service/impl/DataAnonymisationServiceImplTest.java (hunk 2 above)
Comment (Ben-Edwards-cgi) [reply src/test/java/uk/gov/hmcts/darts/common/service/impl/DataAnonymisationServiceImplTest.java]: Updated

File: src/test/java/uk/gov/hmcts/darts/common/service/impl/DataAnonymisationServiceImplTest.java (hunk 3 above)
Comment (Ben-Edwards-cgi) [reply src/test/java/uk/gov/hmcts/darts/common/service/impl/DataAnonymisationServiceImplTest.java]: Updated

File: src/test/java/uk/gov/hmcts/darts/common/service/impl/DataAnonymisationServiceImplTest.java (hunk 6 above)
Comment (Ben-Edwards-cgi) [reply src/test/java/uk/gov/hmcts/darts/common/service/impl/DataAnonymisationServiceImplTest.java]: Updated

File: src/test/java/uk/gov/hmcts/darts/common/service/impl/DataAnonymisationServiceImplTest.java (hunk 5 above)
Comment (Ben-Edwards-cgi) [reply src/test/java/uk/gov/hmcts/darts/common/service/impl/DataAnonymisationServiceImplTest.java]: Updated

File: src/test/java/uk/gov/hmcts/darts/common/service/impl/DataAnonymisationServiceImplTest.java (hunk 7 above)
Comment (Ben-Edwards-cgi) [reply src/test/java/uk/gov/hmcts/darts/common/service/impl/DataAnonymisationServiceImplTest.java]: Updated

File: src/test/java/uk/gov/hmcts/darts/common/service/impl/DataAnonymisationServiceImplTest.java (hunk 8 above)
Comment (Ben-Edwards-cgi) [reply src/test/java/uk/gov/hmcts/darts/common/service/impl/DataAnonymisationServiceImplTest.java]: Updated

**Review by davet1985 — APPROVED** [review]

**Review by hemantasharma1129 — APPROVED** [review]

**Pull request merged**